src/client.rs

```
@@ -38,18 +38,36 @@ impl PowerClient {
 
 impl Power for PowerClient {
     fn performance(&mut self) -> Result<(), String> {
-        println!("setting power profile to performance");
-        self.call_method::<bool>("Performance", None).map(|_| ())
+        let profile = self.get_profile().ok();
+        let profile = profile.as_ref().map_or("?", |s| s.as_str());
+        if profile != "Performance" {
+            println!("setting power profile to performance");
+            self.call_method::<bool>("Performance", None).map(|_| ())
+        } else {
+            Err(String::from("power profile is already set to performance"))
+        }
     }
 
     fn balanced(&mut self) -> Result<(), String> {
-        println!("setting power profile to balanced");
-        self.call_method::<bool>("Balanced", None).map(|_| ())
+        let profile = self.get_profile().ok();
+        let profile = profile.as_ref().map_or("?", |s| s.as_str());
+        if profile != "Balanced" {
+            println!("setting power profile to balanced");
+            self.call_method::<bool>("Balanced", None).map(|_| ())
+        } else {
+            Err(String::from("power profile is already set to balanced"))
+        }
     }
 
     fn battery(&mut self) -> Result<(), String> {
-        println!("setting power profile to battery");
-        self.call_method::<bool>("Battery", None).map(|_| ())
+        let profile = self.get_profile().ok();
+        let profile = profile.as_ref().map_or("?", |s| s.as_str());
+        if profile != "Battery" {
+            println!("setting power profile to battery");
+            self.call_method::<bool>("Battery", None).map(|_| ())
+        } else {
+            Err(String::from("power profile is already set to battery"))
+        }
```

We could condense these down to:

```rust
fn balanced(&mut self) -> Result<(), String> {
    self.profile("Balanced")
}

....

fn profile(&mut self, profile: &str) -> Result<(), String> {
    let current = self.get_profile().ok();
    let current = profile.as_ref().map_or("?", |s| s.as_str());
    if current != profile {
        println!("setting power profile to {}", profile);
        self.call_method::<bool>(profile, None).map(|_| ())
    else {
        Err(format!("power profile is already set to {}", profile))
    }
}
```

Although we may not want to return an error for a profile that's already set. Returning `Ok(())` should be fine.

---

Sounds good ill make the changes